src/net_processing.cpp

```diff
@@ -1315,13 +1315,17 @@ void PeerManager::UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockInde
         }
     }
 
-    // Relay to all peers
-    m_connman.ForEachNode([&vHashes](CNode* pnode) {
-        LOCK(pnode->cs_inventory);
-        for (const uint256& hash : reverse_iterate(vHashes)) {
-            pnode->vBlockHashesToAnnounce.push_back(hash);
+    {
+        LOCK(m_peer_mutex);
+        for (auto& it : m_peer_map) {
```

Comment:
a50dee50fd2aee93b41bfb3b04b24939d0c2f6c1 : `ForEachNode` performs an additional check `NodeFullyConnected` (`pnode->fSuccessfullyConnected && !pnode->fDisconnect;`).

Not sure if that matters much here. Could we prematurely send headers before `VERACK`?

Reply:
This is an excellent observation. Block hashes could be pushed into `m_blocks_for_headers_relay` before the version/verack handshake is complete. However, we'll never send any blocks inventory before we're fully connected due to the guard clause at the top of SendMessages():

https://github.com/bitcoin/bitcoin/blob/c434e2cca9181ce184efef946ab2ce62f2cd2ee7/src/net_processing.cpp#L4077-L4078

I think this behaviour is ok. The version-verack handshake usually lasts no more than a second or two, and at the worst case times out after a minute, so we're never going to push many block hashes onto this vector.